### PR TITLE
add gifting emails - with dummy data

### PR DESCRIPTION
**Pull request by johnduffell**
## Why are you doing this?

As part of the DS gifting work, we need to trigger a pair of emails at purchase time (thank you email and giftee notification email) and a separate one at redemption time for the giftee when they redeem.

[**Trello Card**](https://trello.com/c/RvP8A1nw/3192-gifting-email-api-triggers-for-the-gifter-giftee-emails-generated)

## Changes

* make the email builder return a List rather than a single email
* model the attributes as an ADT rather than building a list of key/value pairs (for Digi sub only at the moment)
* add logic to send the three new emails (as placeholder data only)

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -37,6 +37,7 @@ inThisBuild(Seq(
   // https://www.scala-sbt.org/1.x/docs/Cached-Resolution.html
   updateOptions := updateOptions.value.withCachedResolution(true),
   resolvers ++= Seq(Resolver.sonatypeRepo("releases")), // libraries that haven't yet synced to maven central
+  scalacOptions += "-Ypartial-unification",
```
Comment (johnduffell) [inline build.sbt:40]: this is moved to the top level.  It was already in stripe-intent and payment-api.  It is needed to do sequence on Either[String, A] without needing a type alias. 

> Cats relies on improved type inference via the fix for SI-2712, which is not enabled by default.

from https://typelevel.org/cats/

File: support-modules/aws/src/main/scala/com/gu/aws/AwsCloudWatchMetricPut.scala [hunk 2]
```diff
@@ -45,7 +45,7 @@ object AwsCloudWatchMetricPut {
     metricDatum1.setValue(request.value)
     metricDatum1.setUnit(StandardUnit.Count)
     putMetricDataRequest.getMetricData.add(metricDatum1)
-    Try(client.putMetricData(putMetricDataRequest)).map(_ => Unit)
+    Try(client.putMetricData(putMetricDataRequest)).map(_ => ())
```
Comment (johnduffell) [inline support-modules/aws/src/main/scala/com/gu/aws/AwsCloudWatchMetricPut.scala:48]: this was just a bug/nonsense and it doesn't even compile in scala 2.13.  Can put in another PR if necessary but it will need fixing at some point.

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala [hunk 3]
```diff
@@ -1,78 +1,242 @@
 package com.gu.emailservices
```
Comment (johnduffell) [inline support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:1]: this file was originally a single bulder with a couple of branch points to switch between corp/direct.
There is now a top level branching point between corp/direct/gift purchase/gift redemption which then calls out to a separate method for each one.
This is similar in some ways to the SendThankYouEmail buildEmail function, that has a top level branching point to choose between Contribution/DS/paper/weekly and calls out to each builder in a separate file.

I decided not to make a totally different file for each subcategory of DS, as there is some shared code, and putting it in a class rather than an object means we don't have to pass around so many parameters, as they are all in scope.  However I'm happy to keep things consistent in that way if people prefer that.

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala [hunk 4]
```diff
@@ -1,78 +1,242 @@
...
+    subscription_details: String,
+    paymentAttributes: PaymentFieldsAttributes,
+    date_of_first_payment: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifteeNotificationAttributes(
+    gifter_first_name: String,
+    gift_personal_message: String,
+    gift_code: String,
+  ) extends DigitalSubscriptionEmailAttributes
```
Comment (johnduffell) [inline support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:110]: these lines define the interface between support-workers and braze emails in one place.  This should make it easier in future when working out the contract between us and braze.

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala [hunk 5]
```diff
@@ -1,78 +1,242 @@
...
+                  if (pairsSoFar.contains(fieldName))
+                    Left(s"found duplicate key ${fieldName} in case classes")
+                  else
+                    Right(pairsSoFar + (fieldName -> string))
+              }
+          }
+        }
+
+    flattenToPairs(value, Right(Map.empty)).map(_.toList)
   }
```
Comment (johnduffell) [inline support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:222]: this just turns a JsonObject into a List of key/value pairs for insertion into the existing EmailFields object.  We don't want to change the non DS emails at this point, so it makes sense to convert the case classes to the existing structure.

File: support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala [hunk 6]
```diff
@@ -4,13 +4,18 @@ import com.gu.i18n.Currency
...
 import com.gu.support.workers.{BillingPeriod, User}
-import io.circe.generic.auto._
+import io.circe.Encoder
+import io.circe.generic.semiauto._
 import io.circe.syntax._
 
 case class EmailPayloadContactAttributes(SubscriberAttributes: Map[String, String])
 case class EmailPayloadTo(Address: String, ContactAttributes: EmailPayloadContactAttributes)
-case class EmailPayload(To: EmailPayloadTo, DataExtensionName: String, SfContactId: Option[String], IdentityUserId: Option[String]) {
-  lazy val jsonString: String = this.asJson.toString
```
Comment (johnduffell) [inline support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala:13]: this file used to use the "auto" generated encoders, I assume it was because we were serialising based on `this` rather than a compile time known type.  Given that uses shapeless magic, and we don't use them anywhere else, and it wasn't necessary to apply it to `this` anyway, I just converted for consistency.

File: support-workers/src/main/scala/com/gu/support/workers/lambdas/SendThankYouEmail.scala [hunk 7]
```diff
@@ -61,11 +61,11 @@ class SendThankYouEmail(servicesProvider: ServiceProvider = ServiceProvider)
...
           )
-        )
-      case _: DigitalPack => Right(DigitalPackEmailFields.build(
-        getSubscriptionEmailFields(state, directDebitMandateId),
+        ).map(List(_))
+      case d: DigitalPack => new DigitalPackEmailFields(getSubscriptionEmailFields(state, directDebitMandateId)).build(
         paidSubPaymentData = state.paymentOrRedemptionData.left.toOption,
-      ))
+        readerType = d.readerType
+      )
```
Comment (johnduffell) [inline support-workers/src/main/scala/com/gu/support/workers/lambdas/SendThankYouEmail.scala:68]: the builder may return a Left either if we try to send a redemption email for a non corp/gift sub, or if the ADT structure could not be serialised to a flat key/value pairs list.
Because the Either is sequenced this lambda should still be idempodent, so if we want to send multiple emails, and one fails, it will refuse to send either.  This means if it is retrying repeatedy, the gifter wouldn't receive multiple emails!

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 4 above)
Comment (rupertbates) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:110]: I'm also not that sure that all of what we do send is used by the emails so it would be nice to check that before writing a test

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala [hunk 8]
```diff
@@ -1,78 +1,242 @@
...
+object DigitalSubscriptionEmailAttributes {
+
+  implicit val e1: Encoder.AsObject[BasicDSAttributes] = deriveEncoder
+  implicit val e2: Encoder.AsObject[DirectDSAttributes] = deriveEncoder
+  implicit val e3: Encoder.AsObject[GifteeRedemptionAttributes] = deriveEncoder
+  implicit val e4: Encoder.AsObject[GifterPurchaseAttributes] = deriveEncoder
+  implicit val e5: Encoder.AsObject[GifteeNotificationAttributes] = deriveEncoder
+
+  implicit val encoder: Encoder.AsObject[DigitalSubscriptionEmailAttributes] = Encoder.AsObject.instance {
+    case v: BasicDSAttributes => v.asJsonObject
```
Comment (rupertbates) [inline support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:24]: Why encode to JsonObject rather than Json?

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala [hunk 9]
```diff
@@ -1,78 +1,242 @@
...
-    ) ++ fieldsForReaderType
-
-    val dataExtensionName = if (paidSubPaymentData.isDefined) "digipack" else "digipack-corp"
-
-    EmailFields(fields, Left(sfContactId), user.primaryEmailAddress, dataExtensionName)
+    paidSubPaymentData: Option[PaymentMethodWithSchedule],
+    readerType: ReaderType
+  ): Either[String, List[EmailFields]] =
+    (paidSubPaymentData, readerType) match {
+      case (Some(PaymentMethodWithSchedule(pm, paymentSchedule)), ReaderType.Gift) =>
```
Comment (rupertbates) [inline support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:135]: Is there some way to label what each of these cases means? Maybe break out the result into a separate function?
```scala
case (Some(PaymentMethodWithSchedule(pm, paymentSchedule)), ReaderType.Gift) => giftPurchase(pm)
case (Some(PaymentMethodWithSchedule(pm, paymentSchedule)), _) => purchase(pm, paymentSchedule)
```
Not sure if that will just add clutter...

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 5 above)
Comment (rupertbates) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:222]: This is useful information include it in a comment?

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 4 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:110]: agree it would be nice to reassess what is used.  a test would still provide value for regressions until then.

I wouldn't want to remove anything in this refactoring PR as if something breaks it adds uncertainty - in future either @GHaberis could check in braze, or we could do some eyeballing on the emails.  I'm not sure if some of the fields may be stored in braze for analysis purposes?  I think George was creating a doc for the braze->support workers interface but I'm not sure how far he got.

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 5 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:222]: good idea, will do.  I might even pull that function out to a separate file, as it's nothing to do with emails really, it's just a simple conversion function.

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 8 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:24]: more specific type, a Json could also be a JArray, JString, JNumber etc, so it saves checking that before we start traversing in the flattening function.

I'm not totally sold on it (as I prefer json stuff to be as totally standard as possible, as anything new always requires consuilting the docs) so happy to change back and have the extra bit of scala code and a runtime check if you prefer?

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 9 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:135]: yeah actually it's the unclear use of Some/None that's making it hard for me.

another option is to not use "Option" and roll a clearly named type (or even a type alias if that works?), so we would have 
```scala
case (Purchase(purchaseInfo), ReaderType.Gift) => giftPurchase(purchaseInfo.pm)
case (Purchase(purchaseInfo), _) => purchase(purchaseInfo)
case (Redemption, ReaderType.Corporate) => wrap("digipack-corp", buildCorpRedemptionAttributes).map(List(_))
case (Redemption, ReaderType.Gift) => wrap("digipack-gift-redemption", buildGiftRedemptionAttributes).map(List(_))
case (Redemption, _) => Left("can only redeem gift and corporate subs")
```

Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:135]: I'll first try moving the call to "wrap" into the functions that already exist, see how that helps.

Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:135]: turns out it's trivial to alias, seems to work locally
```scala
  def build(
    paidSubPaymentData: Option[PaymentMethodWithSchedule],
    readerType: ReaderType
  ): Either[String, List[EmailFields]] = {

    val Purchase = Some
    val Redemption = None

    (paidSubPaymentData, readerType) match {
      case (Purchase(paymentInfo), ReaderType.Gift) =>
        List(
          buildGiftPurchaseAttributes(paymentInfo.paymentMethod),
          buildGiftNotificationAttributes
        ).sequence
      case (Purchase(paymentInfo), _) => directAttributes(paymentInfo).map(List(_))
      case (Redemption, ReaderType.Corporate) => buildCorpRedemptionAttributes.map(List(_))
      case (Redemption, ReaderType.Gift) => buildGiftRedemptionAttributes.map(List(_))
      case (Redemption, _) => Left("can only redeem gift and corporate subs")
    }
  }
```

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 8 above)
Comment (rupertbates) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:24]: No happy to go with this

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala [hunk 10]
```diff
@@ -1,78 +1,254 @@
...
+    subscription_details = details
+  )
 
-    val dataExtensionName = if (paidSubPaymentData.isDefined) "digipack" else "digipack-corp"
+  def build(
+    paidSubPaymentData: Option[PaymentMethodWithSchedule],
+    readerType: ReaderType
+  ): Either[String, List[EmailFields]] = {
+
+    val Purchase = Some
```
Comment (rupertbates) [inline support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:135]: This is really nice 👏

File: support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala [hunk 11]
```diff
@@ -31,7 +36,7 @@ case class EmailFields(
       DataExtensionName = dataExtensionName,
       SfContactId = userId.left.toOption.map(_.id),
       IdentityUserId = userId.right.toOption.map(_.id)
-    ).jsonString
+    ).asJson.toString
```
Comment (rupertbates) [inline support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala:39]:
```suggestion
    ).asJson.noSpaces
```
toString behaves slightly odd if the Json isn't a string I think and even though this isn't it's probably better for consistency not to use it

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 10 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:135]: while I was looking it up, I found scala 3 has a really middle ground between type alias and value class, which should reduce the boilerplate for impementing a new type just for a bit of validation.
http://dotty.epfl.ch/docs/reference/other-new-features/opaques.html

File: support-workers/src/test/scala/com/gu/support/workers/integration/SendThankYouEmailSpec.scala [hunk 12]
```diff
@@ -122,176 +132,134 @@ object SendContributionEmail extends App {
...
-    ))),
-  )
-  val service = new EmailService(realConfig.contributionThanksQueueName)
-  service.send(ef)
+    )
+  }
 
-}
-object SendWeeklySubscriptionGiftEmail extends App {
+  //
```
Comment (rupertbates) [inline support-workers/src/test/scala/com/gu/support/workers/integration/SendThankYouEmailSpec.scala:233]: Remove comment

File: support-workers/src/test/scala/com/gu/support/workers/integration/SendThankYouEmailSpec.scala [hunk 13]
```diff
@@ -122,176 +132,134 @@ object SendContributionEmail extends App {
...
 
-  val mandateId = "65HK26E"
-  val billingAddressWithCountry = Address(
+  val countryOnlyAddress = Address(lineOne = None, lineTwo = None, city = None, state = None, postCode = None, country = UK)
+
+  val billingOnlyUser = User("1234", addressToSendTo, None, "Mickey", "Mouse", billingAddress = countryOnlyAddress)
+
+  //
```
Comment (rupertbates) [inline support-workers/src/test/scala/com/gu/support/workers/integration/SendThankYouEmailSpec.scala:239]: Remove comment

File: support-workers/src/test/scala/com/gu/support/workers/integration/SendThankYouEmailSpec.scala [hunk 14]
```diff
@@ -122,176 +132,134 @@ object SendContributionEmail extends App {
...
-    Some(new LocalDate(2019, 3, 26)),
-    PaymentMethodWithSchedule(directDebitPaymentMethod, PaymentSchedule(List(Payment(new LocalDate(2019, 3, 25), 37.50)))),
-    giftRecipient = Some(GiftRecipient(None, "Earl", "Palmer", None))
-  )
-
-  val service = new EmailService(realConfig.contributionThanksQueueName)
-  service.send(ef)
 
-}
+  //
```
Comment (rupertbates) [inline support-workers/src/test/scala/com/gu/support/workers/integration/SendThankYouEmailSpec.scala:260]: Remove comment

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 3 above)
Comment (rupertbates) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:1]: Happy to keep it as is

File: support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala (hunk 11 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala:39]: good point, I'll change.
this was actually moved from line 13 above.   I  intended to change to spaces2, as that's what it calls under the covers, and it's more type safe as you mention.
Do you have a particular preference between nospaces and spaces2?

Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala:39]: image

Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/EmailFields.scala]: I've done spaces2 for now, happy to change before merging though

**Review by rupertbates — APPROVED** [review]
👍 great stuff!

File: support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala (hunk 4 above)
Comment (johnduffell) [reply support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala:110]: I have added tests for my own piece of mind - they have shown that there are no changes to the JSON output for corporate and direct digital subs.  From visual looking, there are 3 fields we may be able to remove from DS direct so I may play with that a bit.
image
